File: 3474.lexicographically-smallest-generated-string.py
#
# @lc app=leetcode id=3474 lang=python3
#
# [3474] Lexicographically Smallest Generated String
#

# @lc code=start
class Solution:
    def generateString(self, str1: str, str2: str) -> str:
        n, m = len(str1), len(str2)
        word_len = n + m - 1
        
        # Initialize result array with 'a' (lexicographically smallest)
        result = ['a'] * word_len
        
        # For each position i in str1, check the constraint
        for i in range(n):
            # Extract the substring of length m starting at position i
            if i + m <= word_len:
                current_substring = ''.join(result[i:i+m])
                
                if str1[i] == 'T':
                    # Must equal str2
                    if current_substring != str2:
                        # Try to make it equal to str2
                        for j in range(m):
                            result[i + j] = str2[j]
                else:  # str1[i] == 'F'
                    # Must NOT equal str2
                    if current_substring == str2:
                        # Need to change at least one character
                        # Change the last character to the next character if possible
                        # or previous if it's 'z'
                        pos = i + m - 1
                        if pos < word_len:
                            if result[pos] < 'z':
                                result[pos] = chr(ord(result[pos]) + 1)
                            else:
                                result[pos] = chr(ord(result[pos]) - 1)
                        else:
                            # This shouldn't happen due to initialization
                            return ""
            else:
                # Not enough characters left for substring of length m
                return ""
        
        # Verify the solution satisfies all constraints
        for i in range(n):
            if i + m <= word_len:
                substring = ''.join(result[i:i+m])
                if str1[i] == 'T' and substring != str2:
                    return ""
                if str1[i] == 'F' and substring == str2:
                    return ""
            else:
                return ""
        
        return ''.join(result)
        
# @lc code=end

File: 3474.lexicographically-smallest-generated-string.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string generateString(string str1, string str2) {
        int n = str1.length();
        int m = str2.length();
        int len = n + m - 1;
        string word(len, '?');
        vector<bool> fixed(len, false);

        for (int i = 0; i < n; ++i) {
            if (str1[i] == 'T') {
                for (int j = 0; j < m; ++j) {
                    if (word[i + j] != '?' && word[i + j] != str2[j]) return "";
                    word[i + j] = str2[j];
                    fixed[i + j] = true;
                }
            }
        }

        for (int i = 0; i < len; ++i) {
            if (word[i] == '?') word[i] = 'a';
        }

        auto check = [&](int i) {
            if (i < 0 || i >= n) return true;
            for (int j = 0; j < m; ++j) {
                if (word[i + j] != str2[j]) return true;
            }
            return false;
        };

        for (int i = 0; i < n; ++i) {
            if (str1[i] == 'F' && !check(i)) {
                bool resolved = false;
                for (int j = m - 1; j >= 0; --j) {
                    if (!fixed[i + j]) {
                        char original = word[i + j];
                        for (char c = 'b'; c <= 'z'; ++c) {
                            word[i + j] = c;
                            bool ok = true;
                            for (int k = max(0, i + j - m + 1); k <= min(n - 1, i + j); ++k) {
                                if (str1[k] == 'T' && !(!check(k))) { ok = false; break; }
                            }
                            if (ok) {
                                resolved = true;
                                break;
                            }
                        }
                        if (resolved) break;
                        word[i + j] = original;
                    }
                }
                if (!resolved) return "";
            }
        }

        for (int i = 0; i < n; ++i) {
            if (str1[i] == 'T' && check(i)) return "";
            if (str1[i] == 'F' && !check(i)) return "";
        }

        return word;
    }
};